crossdock/src/http_server.js

```diff
@@ -0,0 +1,211 @@
+
+            // do async call to prepareResponse
+            let promise = new RSVP.Promise((resolve, reject) => {
+                this.prepareResponse(joinTraceRequest.downstream, {span: serverSpan}).then((response) => {
+                    resolve(response);
+                });
+            });
+
+            // then return result of prepareResponse
+            promise.then((response) => {
+                let traceResponse = JSON.stringify(response);
+                console.log('join_trace_response:', traceResponse);
+                res.send(traceResponse);
+            });
+        });
+
+        app.listen(8081, () => {
+            console.log('HTTP server listening on port 8081...');
+        });
+    }
+
+    prepareResponse(downstream: Downstream, context: any): any {
+        return new RSVP.Promise((resolve, reject) => {
+            let observedSpan = this.observeSpan(context);
+            let response: TraceResponse = {
+                span: observedSpan,
+                notImplementedError: ''
+            };
+
+            if (downstream) {
+                this.callDownstream(downstream, context).then((downstreamResponse) => {
+                    response.downstream = downstreamResponse;
+                    resolve(response);
+                });
+            } else {
+                resolve(response);
+            }
+
+        });
+    }
+
+    callDownstream(downstream: Downstream, context: any): any {
+        let transport = downstream.transport;
+        if (transport === constants.TRANSPORT_HTTP) {
+            return this.callDownstreamHTTP(downstream, context);
+        } else if (transport === constants.TRANSPORT_TCHANNEL) {
+            return new RSVP.Promise((resolve, reject) => {
+                resolve({ 'notImplementedError': 'TChannel has not been implemented' });
+            });
+        } else if (transport == constants.TRANSPORT_DUMMY) {
+            return new RSVP.Promise((resolve, reject) => {
+                resolve({ 'notImplementedError': 'Dummy has not been implemented' });
+            });
+        } else {
+            return new RSVP.Promise((resolve, reject) => {
+                resolve({ 'notImplementedError': `Unrecognized transport received: ${transport}` })
+            });
+        }
+    }
+
+    callDownstreamHTTP(downstream: Downstream, context: any): any {
+        return new RSVP.Promise((resolve, reject) => {
+
+            // $FlowIgnore - Honestly don't know why flow compalins about family.
+            dns.lookup(downstream.host, (err, address, family) => {
+                if (err) {
+                    return;
+                }
+
+                let port = parseInt(downstream.port);
+                let downstreamUrl = `http:\/\/${address}:${port}/join_trace`;
+
+                let clientSpan = this._tracer.startSpan('client-span', { childOf: context.span.context() });
+                context.span = clientSpan;
```

Comment:
wouldn't this mess up the server span that the server endpoint might try to call Finish on?